### PR TITLE
Nexus 8419 nuget search index

### PR DESCRIPTION
**Pull request by mrprescott**
https://issues.sonatype.org/browse/NEXUS-8419

### CODE REVIEW

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/search/SearchServiceImpl.java [hunk 1]
```diff
@@ -162,7 +163,7 @@ public void put(final Repository repository, final Component component) {
         assets = Lists.newArrayList(tx.browseAssets(component));
       }
       String json = JsonUtils.merge(componentMetadata(component, assets), JsonUtils.from(additional));
-      client.get().prepareIndex(repository.getName(), TYPE, component.getEntityMetadata().getId().toString())
+      client.get().prepareIndex(repository.getName(), TYPE, identifier)
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/search/SearchServiceImpl.java]: minor, but there will be inevitable conflicts with https://github.com/sonatype/nexus-oss/pull/1211/files#diff-fce50f73841ede70a43ed2519c3addbe

Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/search/SearchServiceImpl.java]: Thanks for the heads up.

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java [hunk 2]
```diff
@@ -340,6 +343,13 @@ public void put(final InputStream inputStream) throws IOException, NugetPackageE
         getEventBus().post(new ComponentUpdatedEvent(component, getRepository()));
       }
     }
+
+    if (componentId != null) {
+      try (StorageTx storageTx = openStorageTx()) {
+        maintainAggregateInfo(storageTx, componentId);
+        storageTx.commit();
```
Comment (kellyrob99) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: doesn't this put the component and metadata aggregate updates in separate commits now? I would think they should both be in the same commit, if possible, or at least ensure that they're both committed in the same transaction

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java [hunk 3]
```diff
@@ -389,9 +399,13 @@ public boolean delete(final String id, final String version) throws IOException
...
         return false;
       }
+      final SearchFacet searchFacet = getRepository().facet(SearchFacet.class);
+      final SearchItemId searchId = searchFacet.identifier(component);
       tx.deleteComponent(component);
-      getRepository().facet(SearchFacet.class).delete(component);
       tx.commit();
+      searchFacet.delete(searchId);
+
+      searchFacet.delete(searchId);
```
Comment (kellyrob99) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: you REALLY want this thing deleted, don't you? :)

**Comment by adreghiciu** [comment]
Can you please explain the need for SearchItemId?

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java (hunk 3 above)
Comment (mrprescott) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: ack

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java (hunk 2 above)
Comment (mrprescott) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: Yes, this is temporary; for a reason I don't yet understand the aggregate updates don't 'take' when they're in the same transaction. I suspect some simple but subtle wrinkle somewhere, I'm looking now.

**Comment by mrprescott** [comment]
@adreghiciu I wanted the logic of how to make the index keys in one spot, so format implementers didn't all choose a different scheme, breaking uniqueness, which is the reason for searchFacet.identifier(Component).  I originally had it returning String, but then that lets people bypass .identifier() calls and just pass in anything; type safety for the identifier seemed like a simple enough way to solve that.

**Comment by mrprescott** [comment]
http://bamboo.s/browse/NX3-OSSF211-3 :white_check_mark: 
(Still WIP for now tho.)

**Comment by mrprescott** [comment]
http://bamboo.s/browse/NX3-OSSF211-5 :white_check_mark:

File: plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java (hunk 2 above)
Comment (kellyrob99) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: Did you ever figure this out @mrprescott ?

Comment (mrprescott) [inline plugins/basic/nexus-repository-nuget/src/main/java/com/sonatype/nexus/repository/nuget/internal/NugetGalleryFacetImpl.java]: I can fix this once the other PR is merged. :)

**Comment by kellyrob99** [comment]
+1

**Pull request merged**